salt/file/xmpp/ejabberd_authentication.py

```diff
@@ -0,0 +1,170 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""External authentication helper for ejabberd.
+
+This makes it possible to have multiple passwords per user (up to
+--max-passwords-per-user), salted and hashed and stored in a read-only file
+(--config).
+
+The format of the file is one password per line. Each line has three columns
+separated by colons: username, domain name, and hashed salted password. See
+https://docs.python.org/3/library/crypt.html#crypt.crypt for how to generate
+values for the last column. For example, if the only user were
+alice@example.com and her password were "password", the file might look like:
+
+alice:example.com:$6$wntdG0RUDD9lt304$06Lf2Lrlqn7rgdTc3VL4lFjUOz3AadknDPEmrpWyFr5Jzkv9lQyyRz7mWYJ/ILnBRHfHbori.X4sR9B5DcKB60
+
+This script tries to avoid disclosing information via timing attacks as much as
+possible, so --max-passwords-per-user is designed so that (in theory) it should
+take the same amount of time to check a password regardless of how many
+passwords a user has configured.
+"""
+# Protocol: https://docs.ejabberd.im/developer/guide/#external
+#
+# When working on this script, keep these things in mind:
+#
+# It's fine if misconfiguration causes an exception at startup (but not after
+# startup) or if a protocol error causes an exception. However, untrusted input
+# (e.g., usernames, server names, or passwords) must not crash the script.
+#
+# Exception or log messages must not leak passwords or other sensitive
+# information, so they generally also shouldn't include any unknown information.
+# E.g., "parse error, unexpected bytes b'foo'" could risk leaking sensitive
+# information.
+
+import argparse
+import collections
+import crypt
+import enum
+import hmac
+import pathlib
+import struct
+import sys
+from typing import Collection, Mapping, Tuple
+
+# Map from (user, server) to that user's crypted passwords.
+_Config = Mapping[Tuple[bytes, bytes], Collection[str]]
+
+
+def _args():
+    parser = argparse.ArgumentParser(
+        description='External authentication helper for ejabberd.')
+    parser.add_argument(
+        '--config',
+        type=pathlib.Path,
+        required=True,
+        help='Authentication config file.',
+    )
+    parser.add_argument(
+        '--max-passwords-per-user',
+        default=25,
+        type=int,
+        help='Max number of passwords per user.',
+    )
+    return parser.parse_args()
+
+
+def _config(
+    config_path: pathlib.Path,
+    *,
+    max_passwords_per_user: int,
+) -> _Config:
+    raw_config = collections.defaultdict(list)
+    with config_path.open(mode='rb') as config_file:
+        for line in config_file:
+            user, server, crypted_password = line.rstrip(b'\n').split(b':')
```

Comment:
I don't think the crypted password can have a colon, but regardless I think limiting the split improves readability:

```suggestion
            user, server, crypted_password = line.rstrip(b'\n').split(b':', 2)
```

Reply:
It can't have a colon, https://manpages.debian.org/bullseye/libcrypt-dev/crypt.3.en.html
> It will be entirely printable ASCII, and will not contain whitespace or the characters ‘:’, ‘;’, ‘*’, ‘!’, or ‘\’.

I'd rather keep this parsing stricter, both to catch errors early and to make it easier to add more fields later if needed. (Though I'm not sure how much I'll care about backwards compatibility unless other people start using it.)